File: Autonomous Mobile Robots Course/Chapter15/Lesson3/Chapter15_Lesson3.cpp
/*
Chapter15_Lesson3.cpp
Dynamic Window Approach (DWA) — minimal C++17 implementation (no ROS required).
Build (example):
  g++ -std=c++17 -O2 Chapter15_Lesson3.cpp -o dwa_demo
*/

#include <cmath>
#include <iostream>
#include <limits>
#include <tuple>
#include <utility>
#include <vector>

struct State {
    double x{0.0};
    double y{0.0};
    double theta{0.0};
    double v{0.0};
    double w{0.0};
};

struct Config {
    // Kinematic limits
    double v_min{-0.2};
    double v_max{ 1.0};
    double w_min{-1.5};
    double w_max{ 1.5};

    // Dynamics
    double a_max{0.8};
    double a_brake{1.0};
    double alpha_max{2.0};

    // Discretization
    double dt{0.1};
    double T{2.0};
    double v_res{0.05};
    double w_res{0.1};

    // Geometry
    double robot_radius{0.3};

    // Weights
    double w_heading{0.4};
    double w_clear{0.4};
    double w_speed{0.2};
};

static double wrap_to_pi(double a) {
    const double pi = 3.14159265358979323846;
    a = std::fmod(a + pi, 2.0 * pi);
    if (a < 0) a += 2.0 * pi;
    return a - pi;
}

static State motion_step(const State& s, double v, double w, double dt) {
    State ns = s;
    ns.x += v * std::cos(s.theta) * dt;
    ns.y += v * std::sin(s.theta) * dt;
    ns.theta += w * dt;
    ns.v = v;
    ns.w = w;
    return ns;
}

static std::tuple<double,double,double,double> dynamic_window(const State& s, const Config& cfg) {
    double v_low_dyn = s.v - cfg.a_brake * cfg.dt;
    double v_high_dyn = s.v + cfg.a_max * cfg.dt;
    double w_low_dyn = s.w - cfg.alpha_max * cfg.dt;
    double w_high_dyn = s.w + cfg.alpha_max * cfg.dt;

    double vL = std::max(cfg.v_min, v_low_dyn);
    double vU = std::min(cfg.v_max, v_high_dyn);
    double wL = std::max(cfg.w_min, w_low_dyn);
    double wU = std::min(cfg.w_max, w_high_dyn);
    return {vL, vU, wL, wU};
}

static std::vector<State> rollout(const State& s0, double v, double w, const Config& cfg) {
    int n = static_cast<int>(cfg.T / cfg.dt);
    std::vector<State> traj;
    traj.reserve(n + 1);
    traj.push_back(s0);
    State s = s0;
    for (int i = 0; i < n; ++i) {
        s = motion_step(s, v, w, cfg.dt);
        traj.push_back(s);
    }
    return traj;
}

static double min_clearance(const std::vector<State>& traj, const std::vector<std::pair<double,double>>& obs) {
    double best = std::numeric_limits<double>::infinity();
    for (const auto& st : traj) {
        for (const auto& o : obs) {
            double dx = st.x - o.first;
            double dy = st.y - o.second;
            double d = std::sqrt(dx*dx + dy*dy);
            if (d < best) best = d;
        }
    }
    return best;
}

static double heading_score(const std::vector<State>& traj, const std::pair<double,double>& goal) {
    const auto& last = traj.back();
    double dir = std::atan2(goal.second - last.y, goal.first - last.x);
    double err = wrap_to_pi(dir - last.theta);
    return std::cos(err); // [-1,1]
}

static double speed_score(double v, const Config& cfg) {
    return (v - cfg.v_min) / std::max(1e-9, (cfg.v_max - cfg.v_min));
}

static bool admissible(double v, double clearance, const Config& cfg) {
    double d_stop = (std::max(0.0, v) * std::max(0.0, v)) / std::max(1e-9, 2.0 * cfg.a_brake);
    return clearance > (d_stop + cfg.robot_radius);
}

static void normalize(std::vector<double>& vals) {
    double lo = std::numeric_limits<double>::infinity();
    double hi = -std::numeric_limits<double>::infinity();
    for (double v : vals) { lo = std::min(lo, v); hi = std::max(hi, v); }
    if (std::fabs(hi - lo) < 1e-12) {
        for (auto& v : vals) v = 0.0;
        return;
    }
    for (auto& v : vals) v = (v - lo) / (hi - lo);
}

struct Candidate {
    double h;
    double c;
    double s;
    double v;
    double w;
    std::vector<State> traj;
};

static std::tuple<double,double,std::vector<State>>
dwa_control(const State& s, const std::pair<double,double>& goal,
            const std::vector<std::pair<double,double>>& obs, const Config& cfg) {

    auto [vL, vU, wL, wU] = dynamic_window(s, cfg);

    std::vector<Candidate> cand;
    for (double v = vL; v <= vU + 1e-12; v += cfg.v_res) {
        for (double w = wL; w <= wU + 1e-12; w += cfg.w_res) {
            auto traj = rollout(s, v, w, cfg);
            double clear = min_clearance(traj, obs) - cfg.robot_radius;
            if (!admissible(v, clear, cfg)) continue;
            Candidate c;
            c.h = heading_score(traj, goal);
            c.c = clear;
            c.s = speed_score(v, cfg);
            c.v = v;
            c.w = w;
            c.traj = std::move(traj);
            cand.push_back(std::move(c));
        }
    }

    if (cand.empty()) {
        return {0.0, 0.0, std::vector<State>{s}};
    }

    std::vector<double> hs, cs, ss;
    hs.reserve(cand.size()); cs.reserve(cand.size()); ss.reserve(cand.size());
    for (const auto& c : cand) { hs.push_back(c.h); cs.push_back(c.c); ss.push_back(c.s); }
    normalize(hs); normalize(cs); normalize(ss);

    double bestJ = -1e18;
    size_t bestIdx = 0;
    for (size_t i = 0; i < cand.size(); ++i) {
        double J = cfg.w_heading * hs[i] + cfg.w_clear * cs[i] + cfg.w_speed * ss[i];
        if (J > bestJ) { bestJ = J; bestIdx = i; }
    }

    return {cand[bestIdx].v, cand[bestIdx].w, cand[bestIdx].traj};
}

int main() {
    Config cfg;
    State s;
    std::pair<double,double> goal{8.0, 6.0};

    std::vector<std::pair<double,double>> obs;
    for (int i = 0; i < 15; ++i) obs.push_back({2.0 + i*(5.0/14.0), 3.0});
    for (int i = 0; i < 12; ++i) obs.push_back({4.0, 1.0 + i*(4.0/11.0)});

    for (int k = 0; k < 600; ++k) {
        auto [v, w, traj] = dwa_control(s, goal, obs, cfg);
        s = motion_step(s, v, w, cfg.dt);
        double dist = std::hypot(s.x - goal.first, s.y - goal.second);
        if (k % 20 == 0) {
            std::cout << "k=" << k << " x=" << s.x << " y=" << s.y << " dist=" << dist << " v=" << v << " w=" << w << "\n";
        }
        if (dist < 0.3) break;
    }
    std::cout << "Final: x=" << s.x << " y=" << s.y << " theta=" << s.theta << "\n";
    return 0;
}

File: Autonomous Mobile Robots Course/Chapter15/Lesson3/Chapter15_Lesson3.m
% Chapter15_Lesson3.m
% Dynamic Window Approach (DWA) — MATLAB reference implementation + Simulink build script.
%
% Usage:
%   Chapter15_Lesson3();
%
% Notes:
% - Obstacles are point obstacles (Mx2). Replace with costmap distance queries in real AMR stacks.
% - The Simulink section programmatically builds a simple model for unicycle + DWA controller.

function Chapter15_Lesson3()
    cfg = dwa_config();

    s = struct('x',0,'y',0,'theta',0,'v',0,'w',0);
    goal = [8; 6];

    obs = [];
    obs = [obs; [linspace(2,7,15)' 3*ones(15,1)]];
    obs = [obs; [4*ones(12,1) linspace(1,5,12)']];

    path = [s.x, s.y];
    for k = 1:600
        [v, w, traj] = dwa_control(s, goal, obs, cfg); %#ok<ASGLU>
        s = motion_step(s, v, w, cfg.dt);
        path(end+1,:) = [s.x, s.y]; %#ok<AGROW>
        if norm([s.x; s.y] - goal) < 0.3
            break;
        end
    end

    fprintf('Final state: x=%.3f y=%.3f theta=%.3f (steps=%d)\n', s.x, s.y, s.theta, size(path,1));

    figure; hold on; grid on; axis equal;
    plot(path(:,1), path(:,2), 'LineWidth', 1.5);
    scatter(obs(:,1), obs(:,2), 20, 'filled');
    scatter(goal(1), goal(2), 80, 'p', 'filled');
    legend('trajectory','obstacles','goal');
    title('DWA demo (MATLAB)');
end

function cfg = dwa_config()
    cfg.v_min = -0.2; cfg.v_max = 1.0;
    cfg.w_min = -1.5; cfg.w_max = 1.5;

    cfg.a_max = 0.8;
    cfg.a_brake = 1.0;
    cfg.alpha_max = 2.0;

    cfg.dt = 0.1;
    cfg.T  = 2.0;
    cfg.v_res = 0.05;
    cfg.w_res = 0.1;

    cfg.robot_radius = 0.3;

    cfg.w_heading = 0.4;
    cfg.w_clear   = 0.4;
    cfg.w_speed   = 0.2;
end

function s2 = motion_step(s, v, w, dt)
    s2 = s;
    s2.x = s.x + v*cos(s.theta)*dt;
    s2.y = s.y + v*sin(s.theta)*dt;
    s2.theta = wrapToPi(s.theta + w*dt);
    s2.v = v;
    s2.w = w;
end

function [v_best, w_best, traj_best] = dwa_control(s, goal, obs, cfg)
    [vL, vU, wL, wU] = dynamic_window(s, cfg);

    cand = struct('h',{},'c',{},'sp',{},'v',{},'w',{},'traj',{});
    idx = 0;

    for v = vL:cfg.v_res:(vU + 1e-12)
        for w = wL:cfg.w_res:(wU + 1e-12)
            traj = rollout(s, v, w, cfg);
            clear = min_clearance(traj, obs) - cfg.robot_radius;
            if ~admissible(v, clear, cfg)
                continue;
            end
            idx = idx + 1;
            cand(idx).h = heading_score(traj, goal); %#ok<AGROW>
            cand(idx).c = clear;
            cand(idx).sp = speed_score(v, cfg);
            cand(idx).v = v;
            cand(idx).w = w;
            cand(idx).traj = traj;
        end
    end

    if isempty(cand)
        v_best = 0; w_best = 0; traj_best = {s};
        return;
    end

    hs = normalize([cand.h]);
    cs = normalize([cand.c]);
    ss = normalize([cand.sp]);

    Jbest = -inf; best = 1;
    for i = 1:numel(cand)
        J = cfg.w_heading*hs(i) + cfg.w_clear*cs(i) + cfg.w_speed*ss(i);
        if J > Jbest
            Jbest = J; best = i;
        end
    end

    v_best = cand(best).v;
    w_best = cand(best).w;
    traj_best = cand(best).traj;
end

function [vL, vU, wL, wU] = dynamic_window(s, cfg)
    v_low_dyn = s.v - cfg.a_brake * cfg.dt;
    v_high_dyn = s.v + cfg.a_max * cfg.dt;
    w_low_dyn = s.w - cfg.alpha_max * cfg.dt;
    w_high_dyn = s.w + cfg.alpha_max * cfg.dt;

    vL = max(cfg.v_min, v_low_dyn);
    vU = min(cfg.v_max, v_high_dyn);
    wL = max(cfg.w_min, w_low_dyn);
    wU = min(cfg.w_max, w_high_dyn);
end

function traj = rollout(s0, v, w, cfg)
    n = floor(cfg.T / cfg.dt);
    traj = cell(n+1,1);
    traj{1} = s0;
    s = s0;
    for i = 1:n
        s = motion_step(s, v, w, cfg.dt);
        traj{i+1} = s;
    end
end

function dmin = min_clearance(traj, obs)
    pts = zeros(numel(traj),2);
    for i = 1:numel(traj)
        pts(i,:) = [traj{i}.x, traj{i}.y];
    end
    % pairwise distances N x M
    D = sqrt((pts(:,1) - obs(:,1)').^2 + (pts(:,2) - obs(:,2)').^2);
    dmin = min(D(:));
end

function h = heading_score(traj, goal)
    last = traj{end};
    dir = atan2(goal(2) - last.y, goal(1) - last.x);
    err = wrapToPi(dir - last.theta);
    h = cos(err);
end

function sp = speed_score(v, cfg)
    sp = (v - cfg.v_min) / max(1e-9, (cfg.v_max - cfg.v_min));
end

function ok = admissible(v, clear, cfg)
    dstop = (max(0,v)^2) / max(1e-9, 2*cfg.a_brake);
    ok = clear > (dstop + cfg.robot_radius);
end

function y = normalize(x)
    lo = min(x); hi = max(x);
    if abs(hi - lo) < 1e-12
        y = zeros(size(x));
    else
        y = (x - lo) ./ (hi - lo);
    end
end

function a = wrapToPi(a)
    a = mod(a + pi, 2*pi) - pi;
end

% -------------------------------------------------------------------------
% Simulink: Build a minimal DWA closed-loop model programmatically
% -------------------------------------------------------------------------
function Chapter15_Lesson3_build_simulink()
    mdl = 'Chapter15_Lesson3_DWA_Simulink';
    if bdIsLoaded(mdl); close_system(mdl,0); end
    new_system(mdl); open_system(mdl);

    % Blocks: Integrator for x,y,theta (unicycle), MATLAB Function for DWA,
    % and a "To Workspace" for logging.
    add_block('simulink/Sources/Constant', [mdl '/Goal'], 'Value','[8;6]');
    add_block('simulink/User-Defined Functions/MATLAB Function', [mdl '/DWA_Controller']);
    add_block('simulink/Continuous/Integrator', [mdl '/Int_x']);
    add_block('simulink/Continuous/Integrator', [mdl '/Int_y']);
    add_block('simulink/Continuous/Integrator', [mdl '/Int_theta']);

    add_block('simulink/Sinks/To Workspace', [mdl '/x_log'], 'VariableName','x_log');
    add_block('simulink/Sinks/To Workspace', [mdl '/y_log'], 'VariableName','y_log');

    % Positioning (rough)
    set_param([mdl '/Goal'], 'Position', [40 40 80 70]);
    set_param([mdl '/DWA_Controller'], 'Position', [140 30 280 90]);
    set_param([mdl '/Int_x'], 'Position', [340 20 370 50]);
    set_param([mdl '/Int_y'], 'Position', [340 70 370 100]);
    set_param([mdl '/Int_theta'], 'Position', [340 120 370 150]);
    set_param([mdl '/x_log'], 'Position', [420 20 480 50]);
    set_param([mdl '/y_log'], 'Position', [420 70 480 100]);

    % Fill MATLAB Function code (simplified DWA: outputs v,w based on x,y,theta)
    mf = Simulink.findBlocks(mdl, 'BlockType', 'MATLABFunction');
    mf = mf{1};
    code = [
        "function [v,w] = f(goal,x,y,theta)" newline ...
        "%#codegen" newline ...
        "% Minimal placeholder: replace with full dwa_control logic if desired." newline ...
        "dx = goal(1) - x; dy = goal(2) - y;" newline ...
        "dir = atan2(dy,dx);" newline ...
        "err = atan2(sin(dir-theta), cos(dir-theta));" newline ...
        "v = min(0.8, sqrt(dx^2+dy^2));" newline ...
        "w = max(-1.0, min(1.0, 2.0*err));" newline ...
        "end" newline
    ];
    set_param([mdl '/DWA_Controller'], 'Script', code);

    % Wiring: Goal to controller, controller to dynamics, states back
    add_line(mdl, 'Goal/1', 'DWA_Controller/1');
    add_line(mdl, 'Int_x/1', 'DWA_Controller/2');
    add_line(mdl, 'Int_y/1', 'DWA_Controller/3');
    add_line(mdl, 'Int_theta/1', 'DWA_Controller/4');

    % Unicycle dynamics: xdot=v*cos(theta), ydot=v*sin(theta), thetadot=w
    % Use Gain blocks and Trigonometric function blocks for a complete model
    % (omitted for brevity). Students can complete this as an exercise.

    save_system(mdl);
    fprintf('Built Simulink model: %s\n', mdl);
end
